docs/developers/library_examples.md

```diff
@@ -0,0 +1,960 @@
+            // Do something with it.
+        }
+    }
+    catch (const fs::filesystem_error& e) {
+        // Handle error.
+    }
+}
+```
+
+### iostreams
+
+Available Since: v4.2.6
+
+Demonstrates how to use `dstream` and `default_transport` to read and write data objects.
+
+```c++
+// Defines 3 classes:
+// - idstream: Input-only stream for reading data objects.
+// - odstream: Output-only stream for writing data objects.
+// - dstream : Bidirectional stream for reading and writing data objects.
+#include <irods/dstream.hpp>
+
+// Defines the default transport mechanism for transporting bytes via the iRODS protocol.
+//
+// If you are writing server-side code and wish to enable the server-side API, you must
+// define the following macro before including the transport library following this comment.
+//
+//     IRODS_IO_TRANSPORT_ENABLE_SERVER_SIDE_API
+//
+#include <irods/transport/default_transport.hpp>
+
+void write_to_data_object()
+{
+    // IMPORTANT!!!
+    // ~~~~~~~~~~~~
+    // Notice that this library exists under the "experimental" namespace.
+    // This is important if you're considering using this library. It means that any
+    // library under this namespace could change in the future. Changes are likely
+    // to only occur based on feedback from the community.
+    namespace io = irods::experimental::io;
+
+    auto conn = // Our iRODS connection.
+
+    // Instantiates a new transport object which uses the iRODS protocol to read and
+    // write bytes into a data object. Transport objects are designed to be used by IOStreams
+    // objects such as dstream. "default_transport" is a wrapper around the iRODS C API for
+    // reading and writing data objects.
+    //
+    // You can add support for more transport protocols by implementing the following interface:
+    //
+    //     https://github.com/irods/irods/blob/master/lib/core/include/transport/transport.hpp
+    //
+    io::client::default_transport xport{conn};
+
+    // Here, we are creating a new output stream for writing. If the data object exists, then
+    // the existing data object is opened, else a new data object is created.
+    // We could have also used "dstream" itself, but then we'd need to pass in openmode flags
+    // to instruct iRODS on how to open the data object.
+    io::odstream out{xport, "/path/to/data_object"};
+
+    if (!out) {
+        // Handle error.
+    }
+
+    std::array<char, 4_Mb> buffer{}; // Buffer full of data.
+
+    // This is the fastest way to write data into iRODS via the new stream API.
+    // Bytes written this way are stored directly in the buffer as is.
+    out.write(buffer.data(), buffer.size());
+
+    // This will also write data into the data object. This is slower than the previous method
+    // because stream operators format data.
+    out << "Here is some more data ...\n";
+}
+
+void read_from_data_object()
+{
+    namespace io = irods::experimental::io;
+
+    auto conn = // Our iRODS connection.
```

Comment:
nothing on right hand side?

Reply:
That is correct. The examples are not meant to be runnable as they are. For some things, it is on the reader to fill in gaps.

The library examples also contain examples for connecting to an iRODS server so it shouldn't be difficult for the reader to add that.

Reply:
Maybe we can have `conn` be passed in like in the `print_all_resource_names` example under `query`?

Reply:
Very good idea. Will change.

Reply:
Done.